### PR TITLE
Get logs

### LINKED ISSUE #937: eth_getLogs doesn't work
Hi there,

eth_getLogs returns an error

#### Expected behaviour
```
curl --data '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address":"0x1fd169a4f5c59acf79d0fd5d91d1201ef1bce9f1","fromBlock":"0x0","toBlock":"0x0","topics":[["0x03995a801b13c36325306deef859ef977ce61c6e15a794281bf969d204825227","0x2d105ebbc222c190059b3979356e13469f6a29a350add74ac3bf4f22f16301d6","0x29bf0061f2faa9daa482f061b116195432d435536d8af4ae6b3c5dd78223679b","0x3f6fc303a82367bb4947244ba21c569a5ed2e870610f1a693366142309d7cbea","0x667cb7a1818eacd2e3a421e734429ba9c4c7dec85e578e098b6d72cd2bfbf2f6","0xa749dd3df92cae4d106b3eadf60c0dffd3698de09c67ce58ce6f5d02eb821313","0xde7b64a369e10562cc2e71f0f1f944eaf144b75fead6ecb51fac9c4dd6934885"]]}],"id":64690}' -H "Content-Type: application/json" -X POST https://mainnet.infura.io/
{"jsonrpc":"2.0","id":64690,"result":[]}
```

#### Actual behaviour
```
curl --data '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address":"0x1fd169a4f5c59acf79d0fd5d91d1201ef1bce9f1","fromBlock":"0x0","toBlock":"0x0","topics":[["0x03995a801b13c36325306deef859ef977ce61c6e15a794281bf969d204825227","0x2d105ebbc222c190059b3979356e13469f6a29a350add74ac3bf4f22f16301d6","0x29bf0061f2faa9daa482f061b116195432d435536d8af4ae6b3c5dd78223679b","0x3f6fc303a82367bb4947244ba21c569a5ed2e870610f1a693366142309d7cbea","0x667cb7a1818eacd2e3a421e734429ba9c4c7dec85e578e098b6d72cd2bfbf2f6","0xa749dd3df92cae4d106b3eadf60c0dffd3698de09c67ce58ce6f5d02eb821313","0xde7b64a369e10562cc2e71f0f1f944eaf144b75fead6ecb51fac9c4dd6934885"]]}],"id":64690}' -H "Content-Type: application/json" -X POST http://localhost:8545
{"jsonrpc":"2.0","id":64690,"error":{"code":-32602,"message":"invalid argument 0: json: cannot unmarshal non-string into Go value of type common.Hash"}}
```

### PR DESCRIPTION
**Pull request by JekaMas**
closes #937 

@akme It'd be great if you check the getLogs API.

### CODE REVIEW

**Comment by AlexeyAkhunov** [comment]
Could you merge the master in and get the CI green? I will merge it because it contains some code improvements, and we can address any further issues separarely

File: ethdb/remote/ethbackend.pb.go [hunk 1]
```diff
@@ -1,459 +1,364 @@
...
+var _ = math.Inf
 
-// This is a compile-time assertion that a sufficiently up-to-date version
-// of the legacy proto package is being used.
-const _ = proto.ProtoPackageIsVersion4
+// This is a compile-time assertion to ensure that this generated file
+// is compatible with the proto package it is being compiled against.
+// A compilation error at this line likely means your copy of the
+// proto package needs to be updated.
+const _ = proto.ProtoPackageIsVersion3 // please upgrade the proto package
```
Comment (AlexeyAkhunov) [inline ethdb/remote/ethbackend.pb.go:21]: Is this a downgrade?

Comment (AskAlexSharov) [reply ethdb/remote/ethbackend.pb.go:21]: It depends on what generator version installed locally. 

But i found solution for this: if put exact version to go.mod and install generator by ‘go install’ from app folder (make devtools - does it now) - then it installs version from go.mod file. 

So, if run make devtools in current master and re-gen things, then ProtoPackageIsVersion6 must appear

Comment (JekaMas) [reply ethdb/remote/ethbackend.pb.go:21]: Hmm... I've cleaned module packages and downloaded them once again, and now it generates "SupportPackageIsVersion7". At the end of the day, I updated grpc package and regenerated all files once again. It seems everything is good now and tests are green

**Comment by akme** [comment]
`curl --data '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address":"0x9feb49ac78ec80838665d7ab75e43cfbdc4f4277","fromBlock":"0x0","toBlock":"0x0","topics":[["0x4828028875d55b3eee8f0792f7db619443d485cdf943daabb66aa69b59943d60","0xedfff599186e5df9f7c27e1a4c7ccace9b82f62bf814d7e98b10e07b80d677ce","0x8a0df8ef054fae2c3d2d19a7b322e864870cc9fd3cb07fb9526309c596244bf4","0xb2b3cbe033aa38e93c0eb1ceee79d790e60008cbc021b10eed86e2fa214ae645","0x003195cc3cce6bd8cab724dba36b9fcdbd8b093db6771402b7b0ff3637f79070"]]}],"id":623007}' -H "Content-Type: application/json" -X POST`
```{"jsonrpc":"2.0","id":623007,"error":{"code":-32602,"message":"invalid argument 0: math/big: cannot unmarshal \"\\\"0x0\\\"\" into a *big.Int"}}```

**Comment by AlexeyAkhunov** [comment]
thanks for checking, @akme ! I moved your comment into the issue, because I want to merge this and we can continue fixing in a new PR

File: cmd/rpcdaemon/commands/get_receipts.go [hunk 2]
```diff
@@ -139,3 +234,214 @@ func (api *APIImpl) GetTransactionReceipt(ctx context.Context, hash common.Hash)
...
+	}
+	unfiltered := make([]*types.Log, 0, len(logsList))
+	for _, logs := range logsList {
+		unfiltered = append(unfiltered, logs...)
+	}
+	logs = filterLogs(unfiltered, nil, nil, f.addresses, f.topics)
+	if len(logs) > 0 {
+		// We have matching logs, check if we need to resolve full logs via the light client
+		if logs[0].TxHash == (common.Hash{}) {
+			receipts := rawdb.ReadReceipts(api.dbReader, header.Hash(), header.Number.Uint64(), params.MainnetChainConfig)
```
Comment (AlexeyAkhunov) [inline cmd/rpcdaemon/commands/get_receipts.go:355]: Could you remove the hard-coded config? There is a way to read it from the DB (you can see in the `getReceipt` method or similar)

File: cmd/rpcdaemon/commands/eth_api.go [hunk 3]
```diff
@@ -3,6 +3,7 @@ package commands
 import (
 	"context"
 	"fmt"
+	ethereum "github.com/ledgerwatch/turbo-geth"
```
Comment (AlexeyAkhunov) [inline cmd/rpcdaemon/commands/eth_api.go:6]: Wow, I've never seen that one before, what does it import?

Comment (AlexeyAkhunov) [reply cmd/rpcdaemon/commands/eth_api.go:6]: Found it. It is `interfaces.go`

**Comment by AlexeyAkhunov** [comment]
@JekaMas I think you forgot to include the function `UnmarshalJSON` for your `FilterCriteria`, that is why we cannot parse block numbers. See how it is done in `eth/filters/api.go:483`. There is an alias `FilterCriteria` declared, which is used in the signature of the RPC function, and that alias needs to have `UnmarshalJSON` function to custom parsing

**Comment by JekaMas** [comment]
@akme nice catch, could you try again in the latest commit?

File: cmd/rpcdaemon/commands/get_receipts.go [hunk 4]
```diff
@@ -73,6 +77,101 @@ func (api *APIImpl) GetLogs(ctx context.Context, hash common.Hash) ([][]*types.L
...
+		begin := int64(latest)
+		if crit.FromBlock != nil {
+			begin = crit.FromBlock.Int64()
+		}
+		end := int64(latest)
+		if crit.ToBlock != nil {
+			end = crit.ToBlock.Int64()
+		}
+
+		if begin > end {
```
Comment (AlexeyAkhunov) [inline cmd/rpcdaemon/commands/get_receipts.go:129]: Are you sure we need to error out when being == end or just return empty result?

Comment (AlexeyAkhunov) [reply cmd/rpcdaemon/commands/get_receipts.go:129]: I would check pre-existing behaviour

File: eth/filters/api.go [hunk 5]
```diff
@@ -502,11 +501,11 @@ func (args *FilterCriteria) UnmarshalJSON(data []byte) error {
 		args.BlockHash = raw.BlockHash
 	} else {
 		if raw.FromBlock != nil {
-			args.FromBlock = big.NewInt(raw.FromBlock.Int64())
+			args.FromBlock = rpc.NewRPCBlockNumber(int(raw.FromBlock.Int64()))
```
Comment (AlexeyAkhunov) [inline eth/filters/api.go:504]: Aren't `raw.FromBlock` and `args.FromBlock` of the same type, can you just do assignment?

File: interfaces.go [hunk 6]
```diff
@@ -134,8 +135,8 @@ type ContractCaller interface {
 // FilterQuery contains options for contract log filtering.
 type FilterQuery struct {
 	BlockHash *common.Hash     // used by eth_getLogs, return logs only from block with this hash
-	FromBlock *big.Int         // beginning of the queried range, nil means genesis block
-	ToBlock   *big.Int         // end of the range, nil means latest block
+	FromBlock *rpc.BlockNumber // beginning of the queried range, nil means genesis block
```
Comment (AlexeyAkhunov) [inline interfaces.go:138]: If you are defining `FromBlock` and `ToBlock` to be of these types, perhaps you don't need `UnmarshalJSON` function for `FilterQuery` type anymore? My understanding is that it only existed to convert `*rpc.BlockNumber` to `*big.Int`

Comment (JekaMas) [reply interfaces.go:138]: I still need UnmarshalJSON for `Account` that can be either a single value or a slice

File: eth/filters/api.go (hunk 5 above)
Comment (JekaMas) [reply eth/filters/api.go:504]: Not really, golang doesn't allow `&` operation for type conversion but we have to convert int64 to int or rpc.BlockNumber.

File: interfaces.go (hunk 6 above)
Comment (JekaMas) [reply interfaces.go]: I've reverted this type change

File: cmd/rpcdaemon/commands/get_receipts.go (hunk 4 above)
Comment (JekaMas) [reply cmd/rpcdaemon/commands/get_receipts.go]: As far as I see go-ethereum just ignores this case.

File: cmd/rpcdaemon/commands/get_receipts.go (hunk 2 above)
Comment (JekaMas) [reply cmd/rpcdaemon/commands/get_receipts.go]: Done

File: rpc/types.go [hunk 7]
```diff
@@ -69,6 +69,11 @@ type jsonWriter interface {
 
 type BlockNumber int64
 
+func NewRPCBlockNumber(n int) *BlockNumber {
```
Comment (AlexeyAkhunov) [inline rpc/types.go:72]: Do you still need this function?

File: go.sum [hunk 8]
```diff
@@ -432,8 +432,8 @@ google.golang.org/grpc v1.21.0/go.mod h1:oYelfM1adQP15Ek0mdvEgi9Df8B9CZIaU1084ij
 google.golang.org/grpc v1.23.0/go.mod h1:Y5yQAOtifL1yxbo5wqy6BxZv8vAUGQwXBOALyacEbxg=
 google.golang.org/grpc v1.25.1/go.mod h1:c3i+UQWmh7LiEpx4sFZnkU36qjEYZ0imhYfXVyQciAY=
 google.golang.org/grpc v1.27.0/go.mod h1:qbnxyOmOxrQa7FizSgH+ReBfzJrCY1pSN7KXBS8abTk=
-google.golang.org/grpc v1.30.1 h1:oJTcovwKSu7V3TaBKd0/AXOuJVHjTdGTutbMHIOgVEQ=
-google.golang.org/grpc v1.30.1/go.mod h1:N36X2cJ7JwdamYAgDz+s+rVMFjt3numwzf/HckM8pak=
+google.golang.org/grpc v1.33.0-dev h1:c0EY3sGPLj50wEdGQDpiS3zvk/zdduzrAkJTfa9ocjY=
```
Comment (AlexeyAkhunov) [inline go.sum:435]: Is `dev` an unstable version?

**Comment by akme** [comment]
> @akme nice catch, could you try again in the latest commit?

worked well

File: go.sum (hunk 8 above)
Comment (JekaMas) [reply go.sum:456]: @AskAlexSharov I think only you has the right package and libs versions, I've tried to install versions from master and run `devtools` many times and always get `packageVersion7`.
Could you checkout to this branch and regenerate files from proto files and then push results to the branch?

Comment (AskAlexSharov) [reply go.sum:456]: So. 
1. "Is dev an unstable version?" - 1.31 their current release, 1.32-dev their next release, 1.33-dev their next-next release. 
2. Go code generates by google.golang.org/grpc/cmd/protoc-gen-go-grpc - it means we looking for version reproducibility of this particular tool. 
3. When I run `make devtools` in master - it added line `google.golang.org/grpc/cmd/protoc-gen-go-grpc` to go.mod file (I thought that I added this live with exact version already). Reason why it's not there: `go mod tidy` removes it because our source code doesn't use it - it's true - we don't use this package from source code, but we use it to generate go code. 
4. So, probably we can trick `go mod tidy` - add dummy package where I will use all our binary dependencies. But nope, we can do it only if such dep has non-main package: https://github.com/golang/go/issues/32504
5. But I found how to trick it anyway: https://github.com/ledgerwatch/turbo-geth/pull/1037 
6. I pointed `google.golang.org/grpc/cmd/protoc-gen-go-grpc` to commit of `v1.30.1` tag

**Comment by AskAlexSharov** [comment]
Hi, guys. 
I merged  #1037 to master
then merged master to this branch
then regenerated things with right versions
but need help to test  - because if DB has no receipts it returns empty slice
